core/store/src/trie/shard_tries.rs

```diff
@@ -0,0 +1,326 @@
+        for (hash, value, rc) in deletions.iter() {
+            let storage_rc = storage.retrieve_rc(&hash)?;
+            assert!(*rc <= storage_rc);
+            let key = TrieCachingStorage::get_key_from_shard_id_and_hash(shard_id, hash);
+            if *rc < storage_rc {
+                let bytes = encode_trie_node_with_rc(&value, storage_rc - rc);
+                store_update.set(DBCol::ColState, key.as_ref(), &bytes);
+            } else {
+                store_update.delete(DBCol::ColState, key.as_ref());
+            }
+        }
+        Ok(())
+    }
+
+    fn apply_insertions_inner(
+        insertions: &Vec<(CryptoHash, Vec<u8>, u32)>,
+        tries: ShardTries,
+        shard_id: ShardId,
+        store_update: &mut StoreUpdate,
+    ) -> Result<(), StorageError> {
+        let trie = tries.get_trie_for_shard(shard_id);
+        store_update.tries = Some(tries);
+        let storage = trie.storage.as_caching_storage().expect("Must be caching storage");
+        for (hash, value, rc) in insertions.iter() {
+            let storage_rc = storage.retrieve_rc(&hash)?;
+            let key = TrieCachingStorage::get_key_from_shard_id_and_hash(shard_id, hash);
+            let bytes = encode_trie_node_with_rc(&value, storage_rc + rc);
+            store_update.set(DBCol::ColState, key.as_ref(), &bytes);
+        }
+        Ok(())
+    }
+
+    fn apply_all_inner(
+        trie_changes: &TrieChanges,
+        tries: ShardTries,
+        shard_id: ShardId,
+        apply_deletions: bool,
+    ) -> Result<(StoreUpdate, StateRoot), StorageError> {
+        let mut store_update = StoreUpdate::new_with_trie(tries.clone());
+        ShardTries::apply_insertions_inner(
+            &trie_changes.insertions,
+            tries.clone(),
+            shard_id,
+            &mut store_update,
+        )?;
+        if apply_deletions {
+            ShardTries::apply_deletions_inner(
+                &trie_changes.deletions,
+                tries,
+                shard_id,
+                &mut store_update,
+            )?;
+        }
+        Ok((store_update, trie_changes.new_root))
+    }
+
+    pub fn apply_insertions(
+        &self,
+        trie_changes: &TrieChanges,
+        shard_id: ShardId,
+        store_update: &mut StoreUpdate,
+    ) -> Result<(), StorageError> {
+        ShardTries::apply_insertions_inner(
+            &trie_changes.insertions,
+            self.clone(),
+            shard_id,
+            store_update,
+        )
+    }
+
+    pub fn apply_deletions(
+        &self,
+        trie_changes: &TrieChanges,
+        shard_id: ShardId,
+        store_update: &mut StoreUpdate,
+    ) -> Result<(), StorageError> {
+        ShardTries::apply_deletions_inner(
+            &trie_changes.deletions,
+            self.clone(),
+            shard_id,
+            store_update,
+        )
+    }
+
+    pub fn revert_insertions(
+        &self,
+        trie_changes: &TrieChanges,
+        shard_id: ShardId,
```

Comment:
It is still easy to make a silly yet crucial mistake with API like this if you pass a wrong shard_id. I propose we have ShardTrie and ShardTrieChanges, which wrap the relevant structures and pass along the shard_id, so we never lose the information about a shard, and thus we don’t need to pass it separately relying on the outer code to maintain the invariant of providing the correct shard_id.

I am a huge fun of making the invalid state to be impossible to express, there is a cool talk about this mindset if I can recommend: https://youtu.be/3Q2hQfYW-XM (A hammer you can only hold by the handle)

Reply:
`revert_insertions` and `apply_deletions` are only called in one place by gc.
`apply_insertions` is only called by `WrappedTrieChanges` which wraps `block_hash`, `tries` and `shard_id`.
`apply_all` is called once in state sync, once in genesis builder and in ~30 places in test code.

For gc usages `ShardTrie` wouldn't have a longer lifetime than creating it for calling this function. I think instead it might make sense to have a `SingleShardTrieForTests` so we don't pass dummy shard in 30 places.